### PR TITLE
Implement Effects (FX) and Ramping

### PR DESCRIPTION
**Pull request by prodestrian**
## Overview
I used PR #13 for reference while working on this, but didn't build on top of that branch as it was going to require a huge refactor anyway.

This abstracts away a lot of the animation code into a new `FX` class. The `Lights` class now has two instances of this, one for the Powercell and one for the Cyclotron. For each instance we just need to define a few basics (speed, effect, colour) and the class will take care of everything else on its own.

At the moment this includes two animations which I named:
- Spinning (sequentially light all the LEDs until we get to the end, eg Powercell)
- Cycling (one LED at a time, eg Afterlife cyclotron)

But these will need to be renamed as I don't think they're very clear.

This doesn't include any LED grouping, direction changes, or brightness changes, those can come later once more of the motion side is completed.

However I have confirmed that both animations are available on both lightstrips.
Simply change the code in `Lights.cpp` to switch effects:
```ino
_powercellFX.setEffect(SPINNING);
_cyclotronFX.setEffect(CYCLING);
```
The main sketch shouldn't be aware of Effects or Speeds. It should only set States and pass this into the `Lights` class (via `setState()`. Then it's up to the `Lights` class to access the configuration, translate it, and pass it to the `FX` instances.

So, the main sketch will call `Lights.setState(FIRING);`. The `setState()` method will then load the Powercell and Cyclotron configuration for the firing State, which should say what Color/Speed/Effect to use for each of the FX instances, and then pass those through.
We may need to add some intelligence to the (yet-to-be-written) setter methods in the FX class, so that if the new Effect is the same as the current one we do nothing, but if it has changed we may need to perfect some sort of reset.
We may also need to work out which transitions between Effects are instantaneous vs ramped.
Afterlife SHUTDOWN is to slow the cyclotron down and switch off (maybe fade out at the same time).
Classic SHUTDOWN might be to fade out instantly, no slowing down.
Rather than adding more configuration, we could just add more Effects that users can select from for their Shutdown States:
- FadeOut (Fades out but keeps the same speed)
- SlowDown (Slows down but keeps the same brightness)
- SlowDownFadeOut (Slows down and fades out)

## Testing
I have added some simple test code to the main sketch which:
1. Starts the cyclotron at a slow speed and ramps up over 3 seconds
2. Reads the state of a pushbutton connected to `D0/GPIO16`
3. If the button is pressed, increase the Cyclotron speed
4. When the button is let go, decrease the Cyclotron speed

NOTE: This requires the installation of the `Bounce2` button debouncing library: 
https://github.com/thomasfredericks/Bounce2

It works very well, it does feel like it's spinning up/down like a motor.

Video demo here: https://youtu.be/QQ9rT53DV7s

## Changes made
* Remove all Hardware configuration as this will need to be defined in `Settings.h` from now on
* Include `Ramp.h` library: https://github.com/siteswapjuggler/RAMP
* Rename `Directions` enum values to avoid conflict with RAMP
* Create new `FX` class to handle animations
* Move animations out of `Lights` class
* Add support for Ramping in `FX` class
* Add cycling/spinning animations to start with
* Set Cyclotron to Cycling (Afterlife) and Powercell to Spinning
* Remove unnecessary/outdated comments
* Detail requirements for `Lights::setState` method
* Add proof of concept of changing Cyclotron speed when firing button is pressed, using Ramping
* Remove unnecessary enums
* Allow `Settings.h` to be required multiple times without conflicts

### CODE REVIEW

**Comment by ajquick** [comment]
> * Spinning (sequentially light all the LEDs until we get to the end, eg Powercell)
> * Cycling (one LED at a time, eg Afterlife cyclotron)

Funnily enough, I was thinking these would be more apt to be the other way around. Spinning would be a dot like Afterlife, Cycling would be the filling up and then going back down. In all honesty, it doesn't really matter what we call them. They could be "patternOne" and then in the HTML config they could be referred to as "1984 Powercell Animation". What we refer to them in the code is purely for our own sanity.

**Comment by ajquick** [comment]
I made some changes that hopefully won't be too breaking to your setup.

* Added all the switch inputs on their respective pins and ACTIVE LOW / HIGH configuration.
* Added switch state checks to the main loop.
* Each change triggers a setState call, though nothing is done beyond that.
* Added further tests where you can flip the enable switch to start up, or flip the enable switch to shut down.
* If the enable switch is enabled and the fire button is pressed it will speed up and increase in brightness until the fire button is released.
* Added brightness control, changes to timing and a different RAMP function for IMO better movement. Just for testing.

For the motion, I think it would be good to see if we can set a stop value or make speed inverted. For example, I think if the speed value is 255, the movement should be stopped and the LED off. Although I think perhaps the speed should be an equation of 255 - SPEED, so that a speed of *zero* is zero speed and a speed of 255 is a maximum speed if that makes sense.

**Review by ajquick — APPROVED** [review]

**Comment by ajquick** [comment]
A pack state of ```OVERHEATING``` will never be triggered by a button press. So I'm assuming it would be a configurable setting with a workflow such as:

1. Start firing.
2. Fire for 10 seconds. (configurable)
3. ``setState(OVERHEATING)``
4. Overheat for 5 seconds (configurable)
5. ``setState(VENTING)``
6. Vent for 5 seconds (configurable)
7. Automatically go back to start up or idle (configurable)

Unless of course overheating is not configured then it won't. Some people may want to go directly from firing to venting automatically. Others may want to be able to fire forever without automatic venting and would only vent manually.

Perhaps this can also work for Startup leading to an idle state.

1. Start up.
2. Start up duration set to 10 seconds (default, but also configurable)
3. ``setState(IDLE)``

etc.

**Comment by ajquick** [comment]
Added two new effects:

* RAINBOW
* RAINBOW_SCROLL

Rainbow (left) sets all LEDs to a single color and then cycles through the rainbow setting all LEDs the same.
Rainbow scroll (right) sets all LEDs to a different color and then increments through the rainbow, giving a scrolling rainbow effect.

Rainbow Light Modes

I'm not sure if we want to handle colors separately from animations however? Like we could apply the powercell animation with a rainbow effect instead of just blue? (This will actually be important for Videogame mode where the animation might not change, but the color definitely will.)

I didn't want to add any new variables, so I just used the counter for the LEDs. Basically the colors revolve based on how many LEDs are in the string. So 40 LEDs means it shows 40 colors rather than 255 different colors.

**Comment by ajquick** [comment]
Added a "Cylon" effect. That is a dot going back and forth and leaving a fading trail of lights behind it. This adds a function called ```_fadeall()``` which can be used for other effects.

Cylon Effect

I also changed the standard Powercell animation to go up and down as expected.

**Comment by ajquick** [comment]
Added a light function called "Alternate" but may be better called "marque". This function lights up all even numbered LEDs and then all odd numbered LEDs. This function is used in the Powercell of the GBFans Powercell kit during overheating for example. Since this is a very basic effect, if it runs too fast, it will just appear as all lights lit up at once. Recommend to run at 100 or slower.

Alternate Effect

Depending on how you focus your eyes, it will give the illusion of moving both clockwise and counter-clockwise at the same time.

**Comment by ajquick** [comment]
My apologies if I should be working on my own branch of this as I've made some bigger changes. I wanted to address how colors and color changes work. As it currently works, we have just a basic CRGB color defined. While working on the rainbow effects, I realized that it would not be good to have functions that only work with one color set for example.

I found that defined palettes are available in the rainbow color pattern for example. I have expanded the code to now accept both CRGB and Colorpalette16. The pattern function hands off the color decision to a separate function that will decide between a static single color and a pattern. Now it is possible to have a power cell or cyclotron effect that is a rainbow instead of a solid color. FastLED has about a dozen or two predefined patterns. 

I am thinking of refactoring this color function into it's own class. It is possible to gracefully transition between colors, so I think I want to add on the class to do that without cluttering up the FX class. 

Rainbow Powercell

**Comment by prodestrian** [comment]
These are all super cool features, but it might be worth holding off on all the fun details until we get more of the abstract code in place, otherwise we're going to have to rewrite most of this. For example, I still need to write code to handle LED grouping and some more of the timing features. It might end up being best to have them in their own classes (along with color functions) so the FX class just needs to be told which to use, it keeps things simpler. WLED seems to have figured this out in a way that you can combine any animation with any color (solid, rainbow, etc).

But I won't know for sure how to structure this until I really dig into it (tomorrow is my next day to work on it). So we should be able to do the same animations as above but with more configurable options.

I'm going to add a "width" parameter, which is how many LEDs to illuminate at once (how wide is a "pixel"). Perhaps it should be 1,2,3 or more LEDs wide. The animation functions should only need to worry about the pixels. If you have a 60-LED strip and your width is "3", then your animation has 20 frames it can cycle through.

I'm also hoping to add a "blur" feature, so you can choose if you want 3 solidly lit LEDs or if the edges should be dimmer. Ideally this should be calculated against your global brightness level too, so if you're at 50% brightness, your middle LED would be 50%, and the edges could be 25%.

There's a lot left to do, it's just really important to not cram too much into each branch yet or it gets difficult to review and test the changes. I'm hoping once we get to the absolute minimum required functionality we can put a pin in it, tag a `v1.0.0` release, and start getting some testers onboard while we work on the next shiny features (`v1.1.0`).

Classic Mode, Afterlife Mode, some basic customizations, Wifi, and full connectivity with the existing GBFans hardware is probably enough for a `v1.0.0` release (you could start shipping hardware at that point).

**Comment by ajquick** [comment]
Yes I think it would be best to arrange each into classes. Even the patterns should be in their own classes I think. You call the "Cycling" class, it initializes and has it's own variables and then a standard call like ```Cycling:: animate()``` is called. Each pattern would have it's own public animate function which works on the Light object, but then you could also have more private functions if needed.

I basically have the ability right now to swap colors on the fly, switch between colors and patterns and use any pattern in place of the solid colors. It would only take a little bit more to get graceful transitions between colors.

I agree the most important thing right now is probably getting the base structure figured out. Passing the configuration class to the light objects for example. Triggering differences in the states and then forcing a switch to the animation or color, timing and transition to the next state.. etc.

I just enjoy making the patterns work. If I can come up with an effective way to do patterns as a class and also set up a color class, I'll try to do that but I don't want to overstep and work on something you are already tackling. Perhaps I should work on the static html files, CSS and JavaScript that would comprise the config pages and OTA ability?

**Comment by prodestrian** [comment]
There's a lot of really good points in your comments above, I'm still reading through them all. Some we'll want to note down for later when we implement overheating/venting etc.

> A pack state of OVERHEATING will never be triggered by a button press. So I'm assuming it would be a configurable setting with a workflow such as _(truncated)_

Absolutely, it would need to be configurable. I'll write a clean way to determine whether the pack is overheating based on firing duration. In my own pack it slowly cools down after you stop firing (similar to the video game), but I believe the existing GBFans kit instantly cools down?

I've pulled your changes in so I can start to get an idea about how to abstract them a little more.

The input code you added looks great, I think that's another good candidate to abstract out so we don't have to deal with the buttons inside the main sketch. It should be enough for me to exactly replicate the GBFans hardware in my test rig now.

To wrap up this PR I'll focus exclusively on the FX code, I won't need to change the main sketch. 

If you want to start on integrating the WiFi/Server code and the ConfigManager that would be pretty cool.

The `ConfigManager` doesn't need to know about the WiFi side of things, it just needs to read the `config.json` file via LittleFS.
The WiFi/Server code (combined in `WebServer.h` potentially?) just needs to handle writing to that file, as per the demo code we have in another branch.
I feel like we might need another class in the middle, maybe `Config.h`, which would just be a model (DTO) to store all the Configuration and potentially take care of the serialization/deserialization.

I _really_ don't like this code, it's bad now but it's only going to get worse if we don't clean it up:

I'll look at some other large Arduino applications to see how we could potentially manage this.

**Comment by ajquick** [comment]
I'll reply more indepth soon, but I've now looked at many examples of the Arduino Json. I think the ugly method is really the way to do it. Lots of people do it that way. An unfortunate C++ trait that all variables need to be initialized at compile time.

**Comment by ajquick** [comment]
>I'll write a clean way to determine whether the pack is overheating based on firing duration. In my own pack it slowly cools down after you stop firing (similar to the video game), but I believe the existing GBFans kit instantly cools down?

Two ways that this works as far as I know.

* You can set the sound board to allow firing forever. Stop firing and it will do a end firing sound.
* Fire with automatic overheating and venting. If you fire for long enough, it will start overheating and then automatically go into a vent cycle.
* * If you stop firing before it gets to the point to trigger the overheating and the venting, then it will stop firing and do the end firing sound.

So basically allow overheating to be turned off. Allow overheating to happen automatically once a set duration has elapsed.

**Comment by prodestrian** [comment]
I have merged `main` into this branch as this now includes the new GBFans Soundboard Control library (see #15). I will now begin working on integrating this to control the FX.

**Comment by prodestrian** [comment]
## Progress Update - 13 March 2023

With the new Control library merged, I was able to get a working proof of concept in place.

A short demo video can be found here:
https://www.youtube.com/watch?v=N7HdEWHgZ1c

In this demo:
1. The "Sound Board" (actually my Arduino Nano) takes 3 seconds to `POWER_ON`. I currently have no animation for this in the code, so the lightkit doesn't show anything until the Sound Board returns an `IDLE` signal (exactly 3 seconds after the switch is flipped).
2. The Cyclotron and Power Cell both show fairly typical Afterlife `IDLE` animations
3. When a `FIRING` signal is sent, the Power Cell and Cyclotron both ramp up to a faster speed
4. Currently, they don't ramp back down to the original speed (the speed change is immediate)
5. After holding Firing for `5000ms`, the `OVERHEAT` state is set
6. This currently blinks the Cyclotron and N-Filter, and sets an alternating animation on the Power Cell
7. Shutting down the pack (`SHUTDOWN`) causes the Power Cell and Cyclotron to fully light up, before gradually fading out over 2-3 seconds. The N-Filter is also switched off instantly (if it happened to be mid-vent during `SHUTDOWN`).

The actual code in the Main Sketch loop currently looks like this:
```c++
void loop()
{
    controls.update();
    lights.update();

    if (controls.isState(STATE_OFF))
    {
        // Powered off
        Serial.println("Off");
        lights.setState(INACTIVE);
    }
    else if (controls.isState(STATE_IDLE))
    {
        // Idling
        Serial.println("Idle");
        lights.setState(IDLE);
    }
    else if (controls.isState(STATE_FIRE_MOVIE))
    {
        // Firing
        if (controls.duration() > 5000)
        {
            // We have been firing for > 5 seconds
            Serial.println("Overheating");
            lights.setState(OVERHEATING);
        }
        else
        {
            Serial.println("Firing");
            lights.setState(FIRING);
        }
    }
    else if (controls.isState(STATE_POWER_DOWN))
    {
        Serial.println("Powering Down");
        lights.setState(SHUTDOWN);
    }
}
```

Aside from adding more states here, this shouldn't grow much larger than it currently is. Additionally, the 5-second overheat won't be here, as this is controlled from the SoundBoard (I just didn't have time to add this to my Nano).

## Potential Bugs
1. During power-on, one of the LEDS on the N-Filter briefly flashes Green. This isn't occurring on the other LED rings.
2. When the Cyclotron blinks during overheating, the first blink only flashes a single LED, subsequent blinks flash all LEDS in the ring. This will need some investigation into why the code isn't immediately changing states.
3. During shutdown, the Power Cell fades out half as fast as the Cyclotron. I'm not sure why, but for now I've set the fadeouts to different values so they're more synchronised.
4. During Overheating, the Power Cell `ALTERNATE` animation seems to lag every 10 or so frames. I can't see any leftover `delay()` calls in the code, so this could be either a performance lag, or a logic issue with the animation timer code.

## Enhancements

We now have the ability to check for the `duration()` of the current State, in case we need to know this from the Pack itself for some reason.

I have added further debouncing which resolved all of the unusual signals I was receiving from my Arduino Nano (such as random `STATE_AUTOMATIC_VENTING` signals during `SHUTDOWN`). The code now checks to confirm that it has received the same signal `50` times in a row before accepting that it's a legitimate value from the Sound Board and not random noise/interference over the ribbon cable. We can lower this value further but I found I was still getting random signals at `10`, I haven't experimented with values in between. Even at `50` it didn't negatively impact the response time on the Firing button, so "quick press" features like Boson Darts should still work fine.

Finally, I rearranged the libraries into subfolders to help keep things more organized.

## Next Steps

I will need to pick this up again next weekend.
If I can get the remaining simple animations in place (including Startup and Venting), I can then move onto the Configuration feature. As it stands this would be a purely Afterlife pack.
Once Configuration is working reliably we can start expanding to include Classic and Video Game modes.

**Comment by prodestrian** [comment]
I believe this is now completed, the framework for FX/Ramping are in place and working smoothly. Once Configuration is completed we can flesh out the remaining FX/Modes.

**Comment by ajquick** [comment]
Thank you! Just checked the demo video and it is looking great!

I believe the errant single LEDs has to do with the buffer having data stuck in it. Is it possible to clear all LEDs prior to a state change?
